### ISSUE TITLE
Configuration cache storing invalid cache and causing error on second build.

### ISSUE THREAD

**Issue by doniwinata0309** [gradle/gradle#16482]
I have this script on my application module of android project:
```
def vmSelectionAction = {
    fileTree(libDir).matching {
    }.visit { details ->
    }
}
def task = tasks.findByName("package${targetName}")
task.doFirst(vmSelectionAction)
```

On the first run it will complete and configuration cache store the cache. 
Then modify something on code and rerun again, the build will failed with this error

Execution failed for task ':app:packageDebug'.
> Could not find method fileTree() for arguments [/Users/doni.winata/gradlecc/app/build/intermediates/merged_native_libs/] on task ':app:packageDebug' of type com.android.build.gradle.tasks.PackageApplication.

Previously, I posted my problem in this thread and it seems a bug from gradle:
https://discuss.gradle.org/t/gradle-configuration-cache-failed-could-not-find-method-filetree-on-second-run/39253

### Expected Behavior
I think it should be failed in the first run and gradle should not store that invalid cache

### Current Behavior
Gradle config cache store the cache, and  failed on second run

### Steps to Reproduce 
[gradlecc.zip](https://github.com/gradle/gradle/files/6112893/gradlecc.zip)
This is android project to repro the issue. Run the project normally, make changes on app module, and rerun the build will repro the error.

**Comment by bamboo** [comment]
Related to https://github.com/gradle/gradle/issues/20126

**Comment by adammurdoch** [comment]
This is fixed in Gradle 8.0. The build will fail in the first run.
